### PR TITLE
API endpoints for discover

### PR DESCRIPTION
**Pull request by henryc132**

### CODE REVIEW

File: Server/App/api/discover.py [hunk 1]
```diff
@@ -1,35 +1,66 @@
...
-
 class TopHandler(BaseHandler):
     @authorize_request_and_create_db_connector
-    def get(self, connector):
+    @require_query_string_params('page')
+    def get(self, page, connector):
         """Retrieve metadata for the most popular decks."""
-        self.abort(404)
+
+        """Retrieve a page of deck's metadata"""
```
Comment (ClaytonPassmore) [inline Server/App/api/discover.py]: Nit: can you change this from a doc string to a regular comment?

File: Server/App/api/discover.py [hunk 2]
```diff
@@ -1,35 +1,66 @@
...
+
+        # Return search results to the user.
+        return self.make_response(response=json.dumps(decks, default=serializer),
+            status=StatusCode.OK)
 
 class SearchHandler(BaseHandler):
     @authorize_request_and_create_db_connector
-    @require_query_string_params('query')
-    def get(self, query, connector):
+    @require_query_string_params('query', 'tags', 'page')
```
Comment (ClaytonPassmore) [inline Server/App/api/discover.py:45]: Would you mind adding any parameters that you added to the API doc?

File: Server/App/api/discover.py [hunk 3]
```diff
@@ -1,35 +1,66 @@
...
+    def get(self, query, tags_json, page, connector):
         """Retrieve metadata for decks matching the search string."""
-        self.abort(404)
 
+	tags = json.loads(tags_json)
+
+        tags_string = ",".join(tags)
```
Comment (ClaytonPassmore) [inline Server/App/api/discover.py]: I think this line is indented one space too far

File: Server/SQL/schema.sql [hunk 4]
```diff
@@ -15,7 +15,7 @@ CREATE TABLE IF NOT EXISTS Deck
 (
     uuid VARCHAR(36) NOT NULL PRIMARY KEY,
     created TIMESTAMP NOT NULL DEFAULT NOW(),
-    last_update TIMESTAMP NOT NULL,
+    last_update TIMESTAMP NULL,
```
Comment (ClaytonPassmore) [inline Server/SQL/schema.sql:18]: Why did you change this?

Comment (henryc132) [reply Server/SQL/schema.sql:18]: Did this line work for you when you built? Its was set as not null, but given no default, so it wouldn't build for me. And since we can't set it to Now() because of limitations, I figured we might as well let it be null

File: Server/SQL/procedures.sql [hunk 5]
```diff
@@ -263,5 +307,34 @@ BEGIN
...
+        SELECT version, last_update_device, deck_id
+        FROM Library
+    ) AS L ON L.deck_id=Deck.uuid
+    WHERE Deck.public=TRUE
+    AND (Deck.name like CONCAT('%',query_string,'%')
+        OR Deck.uuid IN (SELECT dt.deck_id FROM DeckTag dt
+	    LEFT JOIN Tag t ON dt.tag_id=t.id
+	    WHERE FIND_IN_SET(t.tag,tags)
+	    GROUP BY dt.deck_id
+	    HAVING COUNT(dt.tag_id) = num_tags));
```
Comment (ClaytonPassmore) [inline Server/SQL/procedures.sql]: Should we maybe order by rating as well?

File: Server/SQL/procedures.sql [hunk 6]
```diff
@@ -263,5 +307,34 @@ BEGIN
...
+        Deck.created,
+        Deck.last_update,
+        L.last_update_device,
+        Deck.share_code
+    FROM Deck LEFT JOIN (
+        SELECT version, last_update_device, deck_id
+        FROM Library
+    ) AS L ON L.deck_id=Deck.uuid
+    WHERE Deck.public=TRUE
+    AND (Deck.name like CONCAT('%',query_string,'%')
```
Comment (ClaytonPassmore) [inline Server/SQL/procedures.sql]: I think for all string matches we should make it case insensitive

File: Server/SQL/schema.sql (hunk 4 above)
Comment (ClaytonPassmore) [reply Server/SQL/schema.sql:18]: If you use the `CREATE_DECK` procedure, it will fill this in with `NOW()`

File: Server/App/api/discover.py [hunk 7]
```diff
@@ -1,35 +1,66 @@
...
 class TopHandler(BaseHandler):
     @authorize_request_and_create_db_connector
-    def get(self, connector):
-        """Retrieve metadata for the most popular decks."""
-        self.abort(404)
+    @require_query_string_params('page')
+    def get(self, page, connector):
+        #Retrieve metadata for the most popular decks.
+
+        #Retrieve a page of deck's metadata"""
```
Comment (ClaytonPassmore) [inline Server/App/api/discover.py]: Trailing `"""`

File: Server/App/api/discover.py [hunk 8]
```diff
@@ -1,35 +1,66 @@
...
 
-
 class TopHandler(BaseHandler):
     @authorize_request_and_create_db_connector
-    def get(self, connector):
-        """Retrieve metadata for the most popular decks."""
-        self.abort(404)
+    @require_query_string_params('page')
+    def get(self, page, connector):
+        #Retrieve metadata for the most popular decks.
```
Comment (ClaytonPassmore) [inline Server/App/api/discover.py]: This can stay as a doc string

File: Server/App/api/discover.py (hunk 7 above)
Comment (ClaytonPassmore) [reply Server/App/api/discover.py]: Also add a space after the `#`

File: Server/SQL/procedures.sql [hunk 9]
```diff
@@ -263,5 +307,35 @@ BEGIN
...
+        Deck.share_code
+    FROM Deck LEFT JOIN (
+        SELECT version, last_update_device, deck_id
+        FROM Library
+    ) AS L ON L.deck_id=Deck.uuid
+    WHERE Deck.public=TRUE
+    AND (LOWER(Deck.name) like CONCAT('%',LOWER(query_string),'%')
+        OR Deck.uuid IN (SELECT dt.deck_id FROM DeckTag dt
+	    LEFT JOIN Tag t ON dt.tag_id=t.id
+	    WHERE FIND_IN_SET(t.tag,tags)
```
Comment (ClaytonPassmore) [inline Server/SQL/procedures.sql]: I think we need to do lowercase on tags too... Or should we enforce that all tags must always be lowercase so that we don't have to do it here?

Comment (henryc132) [reply Server/SQL/procedures.sql]: I think we enforce that we save it in lower?

Comment (henryc132) [reply Server/SQL/procedures.sql]: But theres no extra cost in using lower here anyways

Comment (ClaytonPassmore) [reply Server/SQL/procedures.sql]: ok!

Comment (ClaytonPassmore) [reply Server/SQL/procedures.sql]: Well find_in_set isn't case insensitive right? So we would have to make things lowercase first... So instead of adding that we can just force all tags to be lowercase like you said.

File: Server/App/api/discover.py [hunk 10]
```diff
@@ -1,35 +1,66 @@
...
+    @require_query_string_params('query', 'tags', 'page')
+    def get(self, query, tags_json, page, connector):
         """Retrieve metadata for decks matching the search string."""
-        self.abort(404)
 
+        tags = json.loads(tags_json)
+
+        tags_string = ",".join(tags)
+
+	decks = connector.call_procedure('SEARCH_DECKS', query, tags_string, len(tags), page)
```
Comment (ClaytonPassmore) [inline Server/App/api/discover.py]: There's some weird indentation going on here.. lines 53, 55, and 59 need to be indented by one more space.

File: Server/SQL/procedures.sql [hunk 11]
```diff
@@ -263,5 +307,35 @@ BEGIN
...
+        Deck.share_code
+    FROM Deck LEFT JOIN (
+        SELECT version, last_update_device, deck_id
+        FROM Library
+    ) AS L ON L.deck_id=Deck.uuid
+    WHERE Deck.public=TRUE
+    AND (LOWER(Deck.name) like CONCAT('%',LOWER(query_string),'%')
+        OR Deck.uuid IN (SELECT dt.deck_id FROM DeckTag dt
+	    LEFT JOIN Tag t ON dt.tag_id=t.id
+	    WHERE FIND_IN_SET(LOWER(t.tag),LOWER(tags))
```
Comment (ClaytonPassmore) [inline Server/SQL/procedures.sql:334]: Can you apply `lower()` to a set of tags?

**Comment by ClaytonPassmore** [comment]
@henryc132 just a couple of minor comments.

File: Server/App/api/discover.py (hunk 10 above)
Comment (henryc132) [reply Server/App/api/discover.py]: Well that is annoying, look perfectly fine on my editor. I'll try and see whats happening

File: Server/SQL/procedures.sql (hunk 11 above)
Comment (henryc132) [reply Server/SQL/procedures.sql:334]: Tags is simply a comma delimited string, so LOWER is fine in this case

Comment (ClaytonPassmore) [reply Server/SQL/procedures.sql:334]: true (y)

File: Server/App/api/discover.py (hunk 10 above)
Comment (ClaytonPassmore) [reply Server/App/api/discover.py]: could be tabs vs spaces?

**Comment by henryc132** [comment]
@ClaytonPassmore Alright, fixed those darn spacing issues

**Review by ClaytonPassmore — APPROVED** [review]

**Comment by ClaytonPassmore** [comment]
@henryc132 LGTM, merge away
